File: python_parser/octree.py
# -*- coding: utf-8 -*-
# python octree implementation
# Code © Spencer Krum June 2011
# Released underl GPLv3 See LICENSE file in this repository
import numpy as np

class Node:
    """
    An Octree Node
    """

    def __init__(self, parent, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit, Ylowerlimit, Zlowerlimit):
        self.parent = parent
        self.Xupperlimit = Xupperlimit
        self.Yupperlimit = Yupperlimit
        self.Zupperlimit = Zupperlimit
        self.Xlowerlimit = Xlowerlimit
        self.Ylowerlimit = Ylowerlimit
        self.Zlowerlimit = Zlowerlimit
        self.Xcenter = (self.Xupperlimit + self.Xlowerlimit) / 2.
        self.Ycenter = (self.Yupperlimit + self.Ylowerlimit) / 2.
        self.Zcenter = (self.Zupperlimit + self.Zlowerlimit) / 2.
        self.childCount = 0
        self.value = []
        self.values = []
        if parent is not None:
            self.depth = parent.depth + 1

    depth = 0
    childCount = None

    parent = None
    value = None
    values = None

    # children
    posXposYposZ = None
    posXposYnegZ = None
    posXnegYposZ = None
    posXnegYnegZ = None
    negXposYposZ = None
    negXposYnegZ = None
    negXnegYposZ = None
    negXnegYnegZ = None

    # position in space
    Xupperlimit = None
    Yupperlimit = None
    Zupperlimit = None

    Xlowerlimit = None
    Ylowerlimit = None
    Zlowerlimit = None

    def getChildren(self):
        """
        helper function to return array of children
        because there is some weird issue where just setting an
        array variable isn't cutting it
        """
        children = [self.posXposYposZ, self.posXposYnegZ, self.posXnegYposZ, self.posXnegYnegZ, self.negXposYposZ,
                    self.negXposYnegZ, self.negXnegYposZ, self.negXnegYnegZ]
        return children

    def isFullNode(self, maxDepth):
        """
        Checks if a node is 'full' - either it is at a certain depth,
        or it has all the children and all the children are full themselves
        """
        if self.depth >= maxDepth:
            return True

        childFlags = [child is not None and child.isFullNode(maxDepth) for child in self.getChildren()]
        return all(childFlags)

    def getFullNodes(self, maxDepth):
        # return self if full OR if leaf
        if self.isFullNode(maxDepth) or all(child is None for child in self.getChildren()):
            return [self]

        # go through the children and ask them to fetch their leaves
        leaves = []
        for child in self.getChildren():
            if child is not None:
                for childLeaf in child.getFullNodes(maxDepth):
                    leaves.append(childLeaf)
        return leaves

    def getNodesOnLevels(self, levels):
        levels[self.depth].append(self.childCount)

        for child in self.getChildren():
            if child is not None:
                child.getNodesOnLevels(levels)

    def getLeafNodes(self):
        # return self if leaf
        if all(child is None for child in self.getChildren()):
            return [self]

        # go through the children and ask them to fetch their leaves
        leaves = []
        for child in self.getChildren():
            if child is not None:
                for childLeaf in child.getLeafNodes():
                    leaves.append(childLeaf)
        return leaves

    def getNodesAtLevel(self, depth):
        if self.depth >= depth:
            return [self]

        leaves = []
        for child in self.getChildren():
            if child is not None:
                for childLeaf in child.getNodesAtLevel(depth):
                    leaves.append(childLeaf)
        return leaves

    def add(self, payload, coord, level):

        self.childCount += 1
        self.values.append((coord, payload))


        if level == 0:
            self.value.append((coord, payload))

        else:
            level -= 1
            # Determine quadrant
            if coord[0] <= self.Xcenter:
                # negX
                if coord[1] <= self.Ycenter:
                    # negY
                    if coord[2] <= self.Zcenter:
                        # negZ
                        Xupperlimit = self.Xcenter
                        Yupperlimit = self.Ycenter
                        Zupperlimit = self.Zcenter
                        Xlowerlimit = self.Xlowerlimit
                        Ylowerlimit = self.Ylowerlimit
                        Zlowerlimit = self.Zlowerlimit
                        if self.negXnegYnegZ is None:
                            self.negXnegYnegZ = Node(self, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                                     Ylowerlimit, Zlowerlimit)
                        self.negXnegYnegZ.add(payload, coord, level)
                    else:
                        # posZ
                        Xupperlimit = self.Xcenter
                        Yupperlimit = self.Ycenter
                        Zupperlimit = self.Zupperlimit
                        Xlowerlimit = self.Xlowerlimit
                        Ylowerlimit = self.Ylowerlimit
                        Zlowerlimit = self.Zcenter
                        if self.negXnegYposZ is None:
                            self.negXnegYposZ = Node(self, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                                     Ylowerlimit, Zlowerlimit)
                        self.negXnegYposZ.add(payload, coord, level)
                else:
                    # posY
                    if coord[2] <= self.Zcenter:
                        # negZ
                        Xupperlimit = self.Xcenter
                        Yupperlimit = self.Yupperlimit
                        Zupperlimit = self.Zcenter
                        Xlowerlimit = self.Xlowerlimit
                        Ylowerlimit = self.Ycenter
                        Zlowerlimit = self.Zlowerlimit
                        if self.negXposYnegZ is None:
                            self.negXposYnegZ = Node(self, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                                     Ylowerlimit, Zlowerlimit)
                        self.negXposYnegZ.add(payload, coord, level)

                    else:
                        # posZ
                        Xupperlimit = self.Xcenter
                        Yupperlimit = self.Yupperlimit
                        Zupperlimit = self.Zupperlimit
                        Xlowerlimit = self.Xlowerlimit
                        Ylowerlimit = self.Ycenter
                        Zlowerlimit = self.Zcenter
                        if self.negXposYposZ is None:
                            self.negXposYposZ = Node(self, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                                     Ylowerlimit, Zlowerlimit)
                        self.negXposYposZ.add(payload, coord, level)


            else:
                # posX
                if coord[1] <= self.Ycenter:
                    # negY
                    if coord[2] <= self.Zcenter:
                        # negZ
                        Xupperlimit = self.Xupperlimit
                        Yupperlimit = self.Ycenter
                        Zupperlimit = self.Zcenter
                        Xlowerlimit = self.Xcenter
                        Ylowerlimit = self.Ylowerlimit
                        Zlowerlimit = self.Zlowerlimit
                        if self.posXnegYnegZ is None:
                            self.posXnegYnegZ = Node(self, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                                     Ylowerlimit, Zlowerlimit)
                        self.posXnegYnegZ.add(payload, coord, level)

                    else:
                        # posZ
                        Xupperlimit = self.Xupperlimit
                        Yupperlimit = self.Ycenter
                        Zupperlimit = self.Zupperlimit
                        Xlowerlimit = self.Xcenter
                        Ylowerlimit = self.Ylowerlimit
                        Zlowerlimit = self.Zcenter
                        if self.posXnegYposZ is None:
                            self.posXnegYposZ = Node(self, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                                     Ylowerlimit, Zlowerlimit)
                        self.posXnegYposZ.add(payload, coord, level)

                else:
                    # posY
                    if coord[2] <= self.Zcenter:
                        # negZ
                        Xupperlimit = self.Xupperlimit
                        Yupperlimit = self.Yupperlimit
                        Zupperlimit = self.Zcenter
                        Xlowerlimit = self.Xcenter
                        Ylowerlimit = self.Ycenter
                        Zlowerlimit = self.Zlowerlimit
                        if self.posXposYnegZ is None:
                            self.posXposYnegZ = Node(self, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                                     Ylowerlimit, Zlowerlimit)
                        self.posXposYnegZ.add(payload, coord, level)

                    else:
                        # posZ
                        Xupperlimit = self.Xupperlimit
                        Yupperlimit = self.Yupperlimit
                        Zupperlimit = self.Zupperlimit
                        Xlowerlimit = self.Xcenter
                        Ylowerlimit = self.Ycenter
                        Zlowerlimit = self.Zcenter
                        if self.posXposYposZ is None:
                            self.posXposYposZ = Node(self, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                                     Ylowerlimit, Zlowerlimit)
                        self.posXposYposZ.add(payload, coord, level)


class Octree:
    """
    Wrapper class holding the root node and the exposed functions
    """

    def __init__(self, Xmax, Ymax, Zmax, Xmin, Ymin, Zmin, root_coords=(0, 0, 0), maxiter=7):
        self.Xmax = Xmax
        self.Ymax = Ymax
        self.Zmax = Xmax
        self.Xmin = Xmin
        self.Ymin = Ymin
        self.Zmin = Zmin
        self.root_coords = root_coords
        self.maxiter = maxiter

        self.root = Node(None, Xmax, Ymax, Zmax, Xmin, Ymin, Zmin)

    def addItem(self, payload, coord):
        """
        Recursively create subnodes until maxiter is reached
        then deposit payload in that node
        """

        self.root.add(payload, coord, self.maxiter)

    def getLeafNodes(self):
        return self.root.getLeafNodes()

    def getNodesWithAvgChildCount(self, maxChildren):
        levels = {}
        for i in range(0, self.maxiter+1):
            levels[i] = []

        self.root.getNodesOnLevels(levels)

        # print "Average node child count:"
        idx = self.maxiter
        for i in range(0, self.maxiter + 1):
            avgChildren = np.average(levels[i])
            # print str(i) + ": " + str(avgChildren)
            if avgChildren <= maxChildren:
                idx = i
                break

        return self.root.getNodesAtLevel(idx), idx

    def getFullNodes(self):
        return self.root.getFullNodes(self.maxiter)

File: web_app/sim/app/services/Octree.js
/**
 * Created by rokoblak on 8/17/17.
 */

class Octree {

    constructor(positions, maxiter) {
        let xMin = Number.MAX_VALUE;
        let xMax = -Number.MAX_VALUE;
        let yMin = Number.MAX_VALUE;
        let yMax = -Number.MAX_VALUE;
        let zMin = Number.MAX_VALUE;
        let zMax = -Number.MAX_VALUE;

        for (let i = 0; i < positions.length/3; i++) {
            let px = positions[i * 3];
            let py = positions[i * 3 + 1];
            let pz = positions[i * 3 + 2];
            if (px < xMin) {
                xMin = px;
            }
            if (px > xMax) {
                xMax = px;
            }
            if (py < yMin) {
                yMin = py;
            }
            if (py > yMax) {
                yMax = py;
            }
            if (pz < zMin) {
                zMin = pz;
            }
            if (pz > zMax) {
                zMax = pz;
            }
        }

        this.maxiter = maxiter;
        this.root = new Node(null, xMax, yMax, zMax, xMin, yMin, zMin);

        for (let i = 0; i < positions.length/3; i++) {
            let px = positions[i * 3];
            let py = positions[i * 3 + 1];
            let pz = positions[i * 3 + 2];
            this.root.add(i, [px, py, pz], this.maxiter)
        }
    }

    getLeafNodes() {
        return this.root.getLeafNodes()
    }

    getNodesAtLevel(depth) {
        return this.root.getNodesAtLevel(depth)
    }

}

class Node {

    constructor(parent, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit, Ylowerlimit, Zlowerlimit) {
        this.parent = parent;
        this.Xupperlimit = Xupperlimit;
        this.Yupperlimit = Yupperlimit;
        this.Zupperlimit = Zupperlimit;
        this.Xlowerlimit = Xlowerlimit;
        this.Ylowerlimit = Ylowerlimit;
        this.Zlowerlimit = Zlowerlimit;
        this.Xcenter = (this.Xupperlimit + this.Xlowerlimit) / 2.;
        this.Ycenter = (this.Yupperlimit + this.Ylowerlimit) / 2.;
        this.Zcenter = (this.Zupperlimit + this.Zlowerlimit) / 2.;
        this.childCount = 0;
        this.value = [];
        this.values = [];
        if (parent !== null) {
            this.depth = parent.depth + 1
        } else {
            this.depth = 0;
        }

        // children
        this.posXposYposZ = null;
        this.posXposYnegZ = null;
        this.posXnegYposZ = null;
        this.posXnegYnegZ = null;
        this.negXposYposZ = null;
        this.negXposYnegZ = null;
        this.negXnegYposZ = null;
        this.negXnegYnegZ = null;
    }

    getChildren() {
        return [this.posXposYposZ, this.posXposYnegZ, this.posXnegYposZ, this.posXnegYnegZ, this.negXposYposZ,
            this.negXposYnegZ, this.negXnegYposZ, this.negXnegYnegZ];
    }

    getLeafNodes() {
        // return self if leaf
        if (this.getChildren().every((element, index, array) => element === null)) {
            return [this]
        }

        let leaves = [];
        let children = this.getChildren();
        for (let i = 0; i < children.length; i++) {
            let child = children[i];
            if (child !== null) {
                let childLeaves = child.getLeafNodes();
                for (let j = 0; j < childLeaves.length; j++) {
                    leaves.push(childLeaves[j])
                }
            }
        }
        return leaves;
    }

    getNodesAtLevel(depth) {
        if (this.depth >= depth) {
            return [this]
        }

        let leaves = [];
        let children = this.getChildren();
        for (let i = 0; i < children.length; i++) {
            let child = children[i];
            if (child !== null) {
                let childLeaves = child.getNodesAtLevel(depth);
                for (let j = 0; j < childLeaves.length; j++) {
                    leaves.push(childLeaves[j])
                }
            }
        }
        return leaves;
    };

    add(payload, coord, level) {
        this.childCount++;
        this.values.push([coord, payload]);

        if (level === 0) {
            this.value.push([coord, payload])
        } else {
            level--;

            // Determine quadrant
            if (coord[0] <= this.Xcenter) {
                // negX
                if (coord[1] <= this.Ycenter) {
                    // negY
                    if (coord[2] <= this.Zcenter) {
                        // negZ
                        let Xupperlimit = this.Xcenter;
                        let Yupperlimit = this.Ycenter;
                        let Zupperlimit = this.Zcenter;
                        let Xlowerlimit = this.Xlowerlimit;
                        let Ylowerlimit = this.Ylowerlimit;
                        let Zlowerlimit = this.Zlowerlimit;
                        if (this.negXnegYnegZ === null) {
                            this.negXnegYnegZ = new Node(this, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                Ylowerlimit, Zlowerlimit);
                        }
                        this.negXnegYnegZ.add(payload, coord, level);
                    } else {
                        // posZ
                        let Xupperlimit = this.Xcenter;
                        let Yupperlimit = this.Ycenter;
                        let Zupperlimit = this.Zupperlimit;
                        let Xlowerlimit = this.Xlowerlimit;
                        let Ylowerlimit = this.Ylowerlimit;
                        let Zlowerlimit = this.Zcenter;
                        if (this.negXnegYposZ === null) {
                            this.negXnegYposZ = new Node(this, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                Ylowerlimit, Zlowerlimit);
                        }
                        this.negXnegYposZ.add(payload, coord, level);
                    }
                } else {
                    // posY
                    if (coord[2] <= this.Zcenter) {
                        // negZ
                        let Xupperlimit = this.Xcenter;
                        let Yupperlimit = this.Yupperlimit;
                        let Zupperlimit = this.Zcenter;
                        let Xlowerlimit = this.Xlowerlimit;
                        let Ylowerlimit = this.Ycenter;
                        let Zlowerlimit = this.Zlowerlimit;
                        if (this.negXposYnegZ === null) {
                            this.negXposYnegZ = new Node(this, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                Ylowerlimit, Zlowerlimit);
                        }
                        this.negXposYnegZ.add(payload, coord, level);
                    } else {
                        // posZ
                        let Xupperlimit = this.Xcenter;
                        let Yupperlimit = this.Yupperlimit;
                        let Zupperlimit = this.Zupperlimit;
                        let Xlowerlimit = this.Xlowerlimit;
                        let Ylowerlimit = this.Ycenter;
                        let Zlowerlimit = this.Zcenter;
                        if (this.negXposYposZ === null) {
                            this.negXposYposZ = new Node(this, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                Ylowerlimit, Zlowerlimit);
                        }
                        this.negXposYposZ.add(payload, coord, level);
                    }
                }
            } else {
                // posX
                if (coord[1] <= this.Ycenter) {
                    // negY
                    if (coord[2] <= this.Zcenter) {
                        // negZ
                        let Xupperlimit = this.Xupperlimit;
                        let Yupperlimit = this.Ycenter;
                        let Zupperlimit = this.Zcenter;
                        let Xlowerlimit = this.Xcenter;
                        let Ylowerlimit = this.Ylowerlimit;
                        let Zlowerlimit = this.Zlowerlimit;
                        if (this.posXnegYnegZ === null) {
                            this.posXnegYnegZ = new Node(this, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                Ylowerlimit, Zlowerlimit);
                        }
                        this.posXnegYnegZ.add(payload, coord, level);
                    } else {
                        // posZ
                        let Xupperlimit = this.Xupperlimit;
                        let Yupperlimit = this.Ycenter;
                        let Zupperlimit = this.Zupperlimit;
                        let Xlowerlimit = this.Xcenter;
                        let Ylowerlimit = this.Ylowerlimit;
                        let Zlowerlimit = this.Zcenter;
                        if (this.posXnegYposZ === null) {
                            this.posXnegYposZ = new Node(this, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                Ylowerlimit, Zlowerlimit);
                        }
                        this.posXnegYposZ.add(payload, coord, level);
                    }
                } else {
                    // posY
                    if (coord[2] <= this.Zcenter) {
                        // negZ
                        let Xupperlimit = this.Xupperlimit;
                        let Yupperlimit = this.Yupperlimit;
                        let Zupperlimit = this.Zcenter;
                        let Xlowerlimit = this.Xcenter;
                        let Ylowerlimit = this.Ycenter;
                        let Zlowerlimit = this.Zlowerlimit;
                        if (this.posXposYnegZ === null) {
                            this.posXposYnegZ = new Node(this, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                Ylowerlimit, Zlowerlimit);
                        }
                        this.posXposYnegZ.add(payload, coord, level);

                    } else {
                        // posZ
                        let Xupperlimit = this.Xupperlimit;
                        let Yupperlimit = this.Yupperlimit;
                        let Zupperlimit = this.Zupperlimit;
                        let Xlowerlimit = this.Xcenter;
                        let Ylowerlimit = this.Ycenter;
                        let Zlowerlimit = this.Zcenter;
                        if (this.posXposYposZ === null) {
                            this.posXposYposZ = new Node(this, Xupperlimit, Yupperlimit, Zupperlimit, Xlowerlimit,
                                Ylowerlimit, Zlowerlimit);
                        }
                        this.posXposYposZ.add(payload, coord, level);
                    }
                }
            }
        }
    }
}